File: 2-add-two-numbers/2-add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        
        
        carry = 0
        flag = 0
        temp = ListNode()
        total = temp
        while l1 != None or l2!= None:
            
            n1 = 0
            n2 = 0
            if l1 != None:
                n1 = l1.val
                l1 = l1.next
            if l2 != None: 
                n2 = l2.val
                l2 = l2.next
        
            
            addition = n1 + n2 + carry
            to_add = 0
            if addition > 9:
                carry = 1
                to_add = int(str(addition)[1])
            else:
                carry = 0
                to_add = int(addition)
            
            if flag == 0:
                flag = 1
                temp = ListNode(to_add)
                total = temp
            else:
                temp.next = ListNode(to_add)
                temp = temp.next
                
            
            
        if carry == 1:
            temp.next = ListNode(val=1)
        return total
            
        
            
            

File: 2-add-two-numbers/2-add-two-numbers.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
  
    carry := 0
    var sumHead *ListNode
    var sumPrev *ListNode
    for l1 != nil || l2 != nil {
        var v1, v2 int
        if l1 != nil {
            v1 = l1.Val
        } 
        if l2 != nil {
            v2 = l2.Val
        }
        
        newNode := &ListNode{}
        newNode.Val = v1 + v2 + carry
        if newNode.Val >= 10 {
            newNode.Val = newNode.Val % 10
            carry = 1
        } else {
            carry = 0
        }
        
        if sumPrev != nil {
            sumPrev.Next = newNode
        }
        
        // set head of sum list
        if sumHead == nil {
            sumHead = newNode
        } 
        
        sumPrev = newNode
        if l1 != nil {
            l1 = l1.Next
        } 
        if l2 != nil {
            l2 = l2.Next
        }
    }    
    
    if carry == 1 {
        newNode := &ListNode{1, nil}
        sumPrev.Next = newNode
    }
    
    return sumHead
}